### PR TITLE
Adding matrix id validation check. Fixes #1520

### LINKED ISSUE #1520: Verify that Matrix ID is valid
This probably means querying the Matrix server API at the given URL, and checking that a user with that name exists.

### PR DESCRIPTION
**Pull request by dessalines**

### CODE REVIEW

File: crates/api/src/local_user.rs [hunk 1]
```diff
@@ -189,6 +194,12 @@ impl Perform for SaveUserSettings {
       }
     }
 
+    if let Some(Some(matrix_user_id)) = &matrix_user_id {
+      if !is_valid_matrix_id(matrix_user_id) {
```
Comment (Nutomic) [inline crates/api/src/local_user.rs:198]: Maybe trim() before checking/storing?

File: crates/utils/src/utils.rs [hunk 2]
```diff
@@ -114,6 +115,10 @@ pub fn is_valid_preferred_username(preferred_username: &str) -> bool {
     && preferred_username.chars().count() <= 20
 }
 
+pub fn is_valid_matrix_id(matrix_id: &str) -> bool {
+  VALID_MATRIX_ID_REGEX.is_match(matrix_id)
+}
```
Comment (Nutomic) [inline crates/utils/src/utils.rs:120]: We could also check that the domain resolves via DNS, and even get the account info via Matrix API, but thats probably too much for now.

File: crates/api/src/local_user.rs (hunk 1 above)
Comment (dessalines) [reply crates/api/src/local_user.rs:198]: I moved that to the regex, using `^ ... $`. A unit test makes sure that any one with spaces won't validate.

File: crates/utils/src/utils.rs (hunk 2 above)
Comment (dessalines) [reply crates/utils/src/utils.rs:121]: Ya I think for things like email, matrix, etc, regex validation is enough. Who knows what networks lemmy will run on.

**Pull request merged**